### PR TITLE
Tejaswis/avoid additional properties fps

### PR DESCRIPTION
**Pull request by tejaswiMinnu**
RPC-Policy-V1-05 rule reported as FPS and was checking additionalProperties from common-types definitions. [Click here](https://github.com/Azure/azure-rest-api-specs-pr/pull/13713/checks?check_run_id=15182206307)
 This PR will address the issue and rule will not check the additionalProperties defined from common-types or other predefined references.

### CODE REVIEW

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-avoid-additional-properties-fps_2023-08-23-23-01.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft.azure/openapi-validator-rulesets",
+      "comment": "Fix an issue where AvoidAdditionalProperties rule reported as FPS and this rule from now on will not check the additionalProperties defined from common-types or other predefined references.",
```
Comment (rkmanda) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-avoid-additional-properties-fps_2023-08-23-23-01.json:5]: >FPS [](http://example.com/codeflow?start=80&length=3)

what is FPS? please avoid acronyms

Comment (rkmanda) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-avoid-additional-properties-fps_2023-08-23-23-01.json:5]: >and [](http://example.com/codeflow?start=84&length=3)

Small sentences are more readable. So you can split this into 2 sentences.

File: packages/rulesets/src/spectral/functions/avoid-additional-properties.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+const avoidAdditionalProperties = (schema: any, _opts: any, paths: any) => {
+  if (schema === null) {
+    return []
+  }
+  const path = paths.path || {}
+  if ("tags" === path[path.length - 1] && "additionalProperties" in schema) {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/avoid-additional-properties.ts:6]: How is this logic addressing the issue with common types having additionalproperties?

File: packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts [hunk 3]
```diff
@@ -173,6 +224,30 @@ test("AvoidAdditionalProperties should find no errors", () => {
...
+          },
+          tenantId: {
+            readOnly: true,
+            format: "uuid",
+            type: "string",
+            description:
+              "The tenant ID of the system assigned identity. This property will only be provided for a system assigned identity.",
+          },
+          type: {
+            $ref: "src/spectral/test/resources/lro-provisioning-state-specified.json#/definitions/PrivateEndpointConnection",
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:246]: >src/spectral/test/resources/lro-provisioning-state-specified.json#/definitions/PrivateEndpointConnection [](http://example.com/codeflow?start=19&length=104)

does this path resolve locally?

File: packages/rulesets/src/spectral/functions/avoid-additional-properties.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    return
+  } else if ("tags" !== path[path.length - 1] && "additionalProperties" in schema) {
+    return [
+      {
+        message: "The use of additionalProperties is not allowed except for user defined tags on tracked resources.",
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/avoid-additional-properties.ts:11]: >ot allowed except for user defined tags on tracked resources." [](http://example.com/codeflow?start=54&length=62)

The error message and documentation needs to change to reflect the change in logic

**Review by rkmanda — CHANGES_REQUESTED** [review]
:clock1:

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-avoid-additional-properties-fps_2023-08-23-23-01.json (hunk 1 above)
Comment (tejaswiMinnu) [reply common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-avoid-additional-properties-fps_2023-08-23-23-01.json:5]: FPS - false positive.

Comment (tejaswiMinnu) [reply common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-avoid-additional-properties-fps_2023-08-23-23-01.json:5]: Modified and split the line into 2.

File: packages/rulesets/src/spectral/functions/avoid-additional-properties.ts (hunk 2 above)
Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/functions/avoid-additional-properties.ts:6]: Logic is same, I will remove this file. I just modified the resolved to false.

File: packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts (hunk 3 above)
Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:247]: It is not giving error as unresolved as I set resolved to false on the condition, its not getting the additional properties from the PrivateEndpointConnection definition.

File: packages/rulesets/src/spectral/az-arm.ts [hunk 5]
```diff
@@ -234,11 +234,11 @@ const ruleset: any = {
 
     // RPC Code: RPC-Policy-V1-05
     AvoidAdditionalProperties: {
-      description: "The use of additionalProperties is not allowed except for user defined tags on tracked resources.",
+      description: "Definitions must not have properties named additionalProperties except for user defined tags or predefined references.",
       severity: "error",
       stagingOnly: true,
       message: "{{description}}",
-      resolved: true,
+      resolved: false,
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:241]: >false [](http://example.com/codeflow?start=16&length=5)

This fix is incorrect. There could be models referencing other non common types and those need to be resolved. The fix should be to not apply the logic just for managedidentities

File: packages/rulesets/src/spectral/functions/avoid-additional-properties.ts (hunk 4 above)
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/avoid-additional-properties.ts]: >defined tags on tracked resources [](http://example.com/codeflow?start=81&length=33)

and for some system types like ManagedServiceIdentity and PrivateEndpointConnection

Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/functions/avoid-additional-properties.ts]: Removed this file

File: packages/rulesets/src/spectral/az-arm.ts (hunk 5 above)
Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/az-arm.ts]: changed back and addressed delegate and userAssignedIdentity Resources

**Pull request merged**